### PR TITLE
Similarity/matrix updates for relation graph

### LINKED ISSUE #393: Update similarity/matrix to support any relation
Currently, similarity/matrix calculates subsumers list based on rdfs:subClassOf relation. We want to support matrix generated using the given input list of relations.

### PR DESCRIPTION
**Pull request by Shalsh23**
Closes #393

### CODE REVIEW

**Comment by Shalsh23** [comment]
Test results are here: https://github.com/phenoscape/phenoscape-kb-services/issues/393

File: src/main/scala/org/phenoscape/kb/Graph.scala [hunk 1]
```diff
@@ -73,44 +77,63 @@ object Graph {
...
-          case (ancestor, ancPairs) =>
-            val termsForAncestor = ancPairs.map(_._1).toSet
-            val values = termsSequence.map(t => if (termsForAncestor(t)) "1" else "0")
-            s"$ancestor,${values.mkString(",")}"
+        val termSubsumerPairs = pairs.map(p => (p._1 -> (p._2, p._3)))
+
+        // creates Map(term -> virtualIRI(relation, subsumer))
+        val baseIRI = "http://purl.org/phenoscape/term/virtual/owl/ObjectSomeValuesFrom"
+        val termToRelSubsumerSeq = termSubsumerPairs.map { case (term, pairs) =>
+          val virtualTermIRI = baseIRI + URLEncoder.encode("<" + pairs._1 + "> <" + pairs._2 + ">", "UTF-8")
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/Graph.scala:124]: Still seems to be making virtual terms when the relation is `subClassOf`.

**Review by balhoff — CHANGES_REQUESTED** [review]
Don't generate virtual terms for `subClassOf` relations. Are you planning to merge this first, then refactor the common functionality to share with the jaccard method?

**Comment by Shalsh23** [comment]
> Don't generate virtual terms for `subClassOf` relations. Are you planning to merge this first, then refactor the common functionality to share with the jaccard method?

Fixed. New test results here: https://github.com/phenoscape/phenoscape-kb-services/issues/393#issuecomment-824264176

Yes, the plan was to merge this and then refactor it in a new PR.

File: src/main/scala/org/phenoscape/kb/Graph.scala [hunk 2]
```diff
@@ -100,44 +102,68 @@ object Graph {
...
-          case (ancestor, ancPairs) =>
-            val termsForAncestor = ancPairs.map(_._1).toSet
-            val values = termsSequence.map(t => if (termsForAncestor(t)) "1" else "0")
-            s"$ancestor,${values.mkString(",")}"
+        val termSubsumerPairs = pairs.map(p => (p._1 -> (p._2, p._3)))
+
+        // creates Map(term -> virtualIRI(relation, subsumer))
+        val baseIRI = "http://purl.org/phenoscape/term/virtual/owl/ObjectSomeValuesFrom"
+        val termToRelSubsumerSeq = termSubsumerPairs.map { case (term, pairs) =>
+          val virtualTermIRI = RelationalTerm(IRI.create(pairs._1), IRI.create(pairs._2)).iri
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/Graph.scala:149]: Don't create a virtual term for subClassOf.

**Review by balhoff — CHANGES_REQUESTED** [review]

File: src/main/scala/org/phenoscape/kb/Graph.scala (hunk 2 above)
Comment (Shalsh23) [reply src/main/scala/org/phenoscape/kb/Graph.scala]: New test results after update: https://github.com/phenoscape/phenoscape-kb-services/issues/393#issuecomment-825078263

**Review by balhoff — APPROVED** [review]

**Pull request merged**